docs/proposals/201812_thanos-remote-receive.md

```diff
@@ -0,0 +1,174 @@
+
+The hash is roughly calculated as follows.
+
+```
+hash(string(tenant_id) + sort(timeseries.labelset).join())
+```
+
+The hashing function used is the same one as used by Prometheus’: [xxHash][xxHash]. Sorting of labels is necessary, in order to ensure that a unique time-series always has the same hash.
+
+### Ingestion hard tenancy
+
+In attempts to build similar systems a common fallacy has been to distribute all load from all tenants of the system onto a single set of ingestion nodes. This makes reasoning and management rather simple, however, has turned out to have stronger downsides than upsides. Companies using Cortex in production and offering it to clients have setup entirely separate clusters for different customers, because their load has caused incidents affecting other clients. In order to allow customers to send large amounts of data at irregular intervals the ingestion infrastructure needs to scale without impacting durability. Scaling the ingestion infrastructure can cause endpoints to not accept data temporarily, however, the write-ahead-log based replication can cope with this, as it backs off and continues sending its data once the ingestion infrastructure successfully processes those requests again.
+
+Hard tenants in the Thanos receiver are configured in a configuration file, changes to this configuration must be orchestrated by a configuration management tool. When a remote write request is received by a Thanos receiver it goes through the list of configured hard tenants. For each hard tenant there is a separate hash ring respective to their ingestion nodes as described in the "Load distribution" section. A hard tenant also has the number of associated receive nodes belonging to it. A remote write request can be initially received by any receiver node, however, will only be dispatched to receiver nodes that correspond to that hard tenant.
+
+A sample of the configuration of tenants and their respective infrastructure:
+
+```
+tenants:
+- match: tenant-a
+  nodes:
+  - tenant-a-1.metrics.local
+- match: “”
+  nodes:
+  - soft-tenants-1.metrics.local
+```
+
+To start, only exact matches of tenant IDs will used to distribute requests to receive nodes. Should it be necessary, more sophisticated mechanisms can be added later. When a request is received, the specified tenant is tested against the configured tenant ID until an exact match is found. If the specified tenant is the empty string, then any tenant is considered a valid match. If no hard tenancy is configured, a tenant will automatically land in a soft tenancy hashring.
+
+```
+                                  Soft tenant hashring
+                                 +-----------------------+
+                                 |                       |
++-----------------+              |  +-----------------+  |
+|                 |              |  |                 |  |
+|  Load Balancer  +-------+      |  | Thanos receiver |  |
```

Comment:
Having this load balancer immdiately proxy to soft/hard receiver would be bit simpler in terms of operational effort maybe?

I wonder if we can set up some envoy for it (it is L7 proxy) that wil grab tenant ID and know if it's hard or not and forward accordingly?

Reply:
The problem with soft (or any) receiver forwarding to hard tenant receiver is that again, you crash the receiver (so push for plenty of tenants) if one hard tenant creates some DOS

Reply:
Yes that would be perfectly possible, I'm just wanted to avoid creating another additional component, and instead merged this functionality into the receiver.

Reply:
Cool, I totally see that - worth to mention that

Reply:
Adding "While the routing functionality could be a separate component, we choose to have it in the receiver to allow for a simpler setup." To the load distribution section.